rbd/mirror_nautilus.go

```diff
@@ -0,0 +1,379 @@
+type MirrorImageInfo struct {
+	Name        string
+	Description string
+	State       MirrorImageState
+	StatusState MirrorImageStatusState
+	GlobalID    string
+	IsPrimary   bool
+	IsUp        bool
+}
+
+// MirrorGetImage returns the MirrorImageInfo for an image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  rbd_mirror_image_get_info(rbd_image_t image,
+//                            rbd_mirror_image_info_t *mirror_image_info,
+//                            size_t info_size)
+func (image *Image) MirrorGetImage() (*MirrorImageInfo, error) {
+	var status C.rbd_mirror_image_status_t
+
+	ret := C.rbd_mirror_image_get_status(image.image, &status, C.sizeof_rbd_mirror_image_status_t)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	return &MirrorImageInfo{
+		Name:        C.GoString(status.name),
+		Description: C.GoString(status.description),
+		State:       MirrorImageState(status.info.state),
+		StatusState: MirrorImageStatusState(status.state),
+		GlobalID:    C.GoString(status.info.global_id),
+		IsPrimary:   bool(status.info.primary),
+		IsUp:        bool(status.up),
+	}, nil
+}
+
+// MirrorImageList returns a MirrorImageInfo for each mirrored image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  int rbd_mirror_image_status_list(rados_ioctx_t io_ctx,
+//  				      									 const char *start_id, size_t max,
+//  				      									 char **image_ids,
+//  				      									 rbd_mirror_image_status_t *images,
+//  				      									 size_t *len)
+func MirrorImageList(ioctx *rados.IOContext) ([]*MirrorImageInfo, error) {
+	imageInfos := make([]*MirrorImageInfo, 0)
+	const cMaxIter C.size_t = 100
+	var startID string
+
+	for {
+		// We need to wrap all the actions within the for loop in a function
+		// in order to ensure that we correctly reclaim all allocated memory
+		// from C at the end of every iteration.
+		ret, done := iterateImageList(ioctx, &imageInfos, &startID, cMaxIter)
+		if ret != 0 {
+			return imageInfos, getError(ret)
+		}
+
+		if done {
+			break
+		}
+	}
+	return imageInfos, nil
+}
+
+func iterateImageList(ioctx *rados.IOContext, imageInfos *[]*MirrorImageInfo, startID *string, cMaxIter C.size_t) (C.int, bool) {
+	cImageIDs := make([]*C.char, cMaxIter)
+	cImageStatus := make([]C.rbd_mirror_image_status_t, cMaxIter)
+	done := false
+
+	var cLen C.size_t
+	ret := C.rbd_mirror_image_status_list(cephIoctx(ioctx), C.CString(*startID),
+		cMaxIter, &cImageIDs[0], &cImageStatus[0], &cLen)
+	if ret != 0 {
+		return ret, done
+	}
+
+	// If the list length is 0 or less than the max size
+	// specified we know we are on the last page of the list,
+	// and we don't need to continue iterating.
+	if cLen < cMaxIter {
+		done = true
+	}
+
+	if cLen == 0 {
+		return C.int(0), done
+	}
+
+	defer func() {
+		C.rbd_mirror_image_status_list_cleanup(&cImageIDs[0], &cImageStatus[0], cLen)
+	}()
```

Comment:
Not a change request, only a note: I only see one `defer` and one `return` after that, so you don't really need another function here. You could just explicitly call `C.rbd_mirror_image_status_list_cleanup()` after filling `imageInfos`.

Reply:
I prefer the pattern of deferring resource cleanup as soon as the resource is created, but good to know!

Reply:
Sure, if you don't need the performance (function call and defer doesn't come for free) then it's the safer way to do it.